File: C++Implementation/MergeSort.cpp
#include <iostream>
using namespace std;

// Function Declarations
int* mergesort(int[], int);
int* merge(int*, int*, int, int);

int main() {

	int myArray[] = {24, 67, 50, 8, 29, 116, 125, 48, 110, 82, 145, 6, 63, 59, 80, 26, 105, 33, 15, 129 };
	int arraySize = size(myArray);
	int* sortedArray; // Create a pointer which will hold first address of sortedArray
	sortedArray = mergesort(myArray, arraySize); // mergesort will return pointer representing first element of sorted array
	for (int i = 0; i < arraySize; i++) {
		cout << sortedArray[i] << " ";
	}
	return 0;
}

// mergesort takes two arguments: the array i.e. address of first element in array and arraySize
int* mergesort(int myArray[], int arraySize) {

	// Base condition. If the array has a single element, then the recursive function will return the single element in array
	if (arraySize < 2) {
		return myArray;
	}
	else {
		int middle = arraySize / 2; // The middle value will tell us where to split the array from.
		// Determine the size of each sublist
		int leftSize = middle;
		int rightSize = arraySize - middle;
		// Use the sizes to construct the sublists of specified lengths
		int *left, *right;
		left = new int[leftSize];
		right = new int[rightSize];
		// Use a for loop to create the different splits
		for (int i = 0; i < leftSize; i++) {
			left[i] = myArray[i];
		}
		for (int i = 0; i < rightSize; i++) {
			right[i] = myArray[i+middle];
		}
		// Utilize recursion and call mergesort using the split arrays.
		int *newLeft, *newRight, *newList;
		newLeft = mergesort(left, leftSize);
		newRight = mergesort(right, rightSize);
		// Call merge which will merge the two arrays to one large array
		newList = merge(newLeft, newRight, leftSize, rightSize);
		return newList;
	}
}

// Merge takes the address of the first element in the left and right subarray. It also takes their respective sizes.
int* merge(int* left, int* right, int leftSize, int rightSize) {
	int *mergedList;
	mergedList = new int[leftSize + rightSize]; // Create a merged list which is the size of the sum of left Array Size and right Array Size
	
	int i = 0, j = 0, k = 0;
	// While the indexes are less than their respective array lengths, add new values to merged list.
	while ((i < leftSize) && (j < rightSize)) {
		// We will add the smallest value between two values from the left subarray and right subarray
		if (left[i] <= right[j]) {
			mergedList[k] = left[i];
			i++;
			k++;
		}
		else {
			mergedList[k] = right[j];
			j++;
			k++;
		}
	}
	// The last two while loops are designed to deal with any values that have been left over
	while (i < leftSize) {
		mergedList[k] = left[i];
		i++;
		k++;
	}
	while (j < rightSize) {
		mergedList[k] = right[j];
		j++;
		k++;
	}
	return mergedList;
}

File: PythonImplementation/MergeSort.py
# This program will demonstrate Merge Sort which is best utilized when sorting Linked Lists
# Unlike other sorting algorithms, Merge Sort has a Time Complexity of O(n log n)
# However, due to the way in which Merge Sort works, it also has Space Complexity of O(n)

def main():
    myList = [24, 67, 50, 8, 29, 116, 125, 48, 110, 82, 145, 6, 63, 59, 80, 26, 105, 33, 15, 129]
    # We will call a function called mergesort which takes unsorted list as argument. This function will return sorted list in O(n log n) time
    mySortedList = mergesort(myList)
    print(mySortedList)

# In order to perform mergesort, the mergesort function utilizes a concept called recursion. Recursion is when a function calls itself and is useful for situations such as this.
def mergesort(a_list):
    # In recursive functions, there is a base condition which determines when function should exit recursion
    # In our case, when the len of the list is less than 2, return the list
    if len(a_list) < 2:
        return a_list
    else:
        middle = int(len(a_list) / 2) # The middle represents a point where we split the list into two sublists
        left = a_list[:middle] # Create the left sublist by splitting on middle index
        right = a_list[middle:] # Create the right sublist by splitting on midde index
        newLeft = mergesort(left) # Recursive call sending left sublist
        newRight = mergesort(right) # Recursive call sending right sublist
        newList = merge(newLeft,newRight) # Merge will merge the two sorted sublists
        return newList

def merge(list_a, list_b):
    # These variables will represent the indexes used to merge the two sublists to one whole list
    i = 0
    j = 0
    mergedList = []
    # As long as the length of the list has not been met, we will compare each value and decide which value to put before
    while i < len(list_a) and j < len(list_b):
        if list_a[i] <= list_b[j]: # If this condition is met, list_a[i] will be put into mergedList and index will be incremented.
            mergedList.append(list_a[i])
            i += 1
        else:
            mergedList.append(list_b[j]) # If condition was not met, list_b[i] will be put into mergedList and index will be incremented
            j += 1
    # This next portion will handle any thing that was left over in list_a or list_b
    # Since the previous while loop exited due to one of the conditions failing, only one of the while loops on the bottom will run.
    while i < len(list_a):
        mergedList.append(list_a[i])
        i += 1
    while j < len(list_b):
        mergedList.append(list_b[j])
        j += 1
    return mergedList

main()